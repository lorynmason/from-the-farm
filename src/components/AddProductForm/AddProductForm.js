import React, { Component } from 'react';
import { connect } from 'react-redux';
import { postItem } from '../../thunks/postItem';
import { fetchUser } from '../../thunks/fetchUser';
import { Link } from 'react-router-dom';

export class AddProductForm extends Component {
  constructor() {
    super();
    this.state = {
      item_id: '',
      description: '',
      price: 0,
      unit: ''
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const price = this.state.price * 100;
    const url = 'https://xpoll-be.herokuapp.com/api/v1';
    this.props.postItem(`${url}/vendor_items`, {...this.state, price}, this.props.user.token);
    this.props.fetchUser(`${url}/users/${this.props.user.id}`, this.props.user.token)
    this.setState({
      item_id: '',
      description: '',
      price: 0,
      unit: ''
    });
  }

  render() {
    const productOptions = this.props.items.map((item) => {
      return <option value={item.id} key={item.id}>{item.name}</option>
    });

    return (
      <section>
        <form className="add-product-form" onChange={this.handleChange} onSubmit={this.handleSubmit}>
          <select name="item_id" id="">
            <option value="">select a product</option>
            {productOptions}
          </select>
          <input type="text" name="description" placeholder="description" />
          <input type="text" placeholder="unit i.e lb" name="unit" />
          $<input type="number" placeholder="price i.e 5.50" name="price" step=".01" />
          <button>Submit</button>
        </form>
        <p><Link to="/profile">Return to Profile</Link></p>
      </section>
    )
  }
}

export const mapStateToProps = (state) => ({
  items: state.items,
  user: state.user
});

export const mapDispatchToProps = (dispatch) => {
  return {
    postItem: (url, item, token) => dispatch(postItem(url, item, token)),
    fetchUser: (url, token) => dispatch(fetchUser(url, token))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddProductForm);