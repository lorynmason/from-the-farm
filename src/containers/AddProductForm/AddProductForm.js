import React, { Component } from 'react';
import { connect } from 'react-redux';
import { postItem } from '../../thunks/postItem';
import { fetchUser } from '../../thunks/fetchUser';
import { fetchVendors } from '../../thunks/fetchVendors';
import { Link } from 'react-router-dom';
import ProductList from '../../components/ProductList/ProductList';
import { Loading } from '../../components/Loading/Loading';

export class AddProductForm extends Component {
  constructor() {
    super();
    this.state = {
      item_id: '',
      description: '',
      price: 0,
      unit: ''
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  handleSubmit = async(e) => {
    e.preventDefault();
    const price = this.state.price * 100;
    const url = 'https://xpoll-be.herokuapp.com/api/v1';
    await this.props.postItem(`${url}/vendor_items`, {...this.state, price}, this.props.user.token);
    this.props.fetchUser(`${url}/users/${this.props.user.id}`, this.props.user.token)
    this.props.fetchVendors(url+'/vendors')
    this.setState({
      item_id: '',
      description: '',
      price: 0,
      unit: ''
    });
  }

  render() {
    const productOptions = this.props.items.map((item) => {
      return <option value={item.id} key={item.id}>{item.name}</option>
    });
    if(this.props.isLoading) {
      return (<Loading />)
    }

    return (
      <section className="add-product">
        <form className="add-product-form" onSubmit={this.handleSubmit} onChange={this.handleChange}>
          <h3>Add to Your Inventory</h3>
          <p>Add a Product <select name="item_id" id="">
            <option value="" onChange={this.handleChange}>select a product</option>
            {productOptions}
          </select></p>
          <p>Add Description <input type="text" name="description" placeholder="description"  onChange={this.handleChange}/></p>
          <p>Add Unit Amount <input type="text" placeholder="unit i.e lb" name="unit" onChange={this.handleChange}/></p>
          <p>Add a Price $<input type="number" placeholder="price i.e 5.50" name="price" onChange={this.handleChange}/></p>
          <button>Submit</button>
          <p><Link to="/profile">Return to Profile</Link></p>
        </form>
        <section className="my-products">
        <ProductList products={this.props.user.products} />
        </section>
      </section>
    )
  }
}

export const mapStateToProps = (state) => ({
  items: state.items,
  user: state.user,
  isLoading: state.isLoading
});

export const mapDispatchToProps = (dispatch) => {
  return {
    postItem: (url, item, token) => dispatch(postItem(url, item, token)),
    fetchUser: (url, token) => dispatch(fetchUser(url, token)),
    fetchVendors: (url) => dispatch(fetchVendors(url))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddProductForm);
