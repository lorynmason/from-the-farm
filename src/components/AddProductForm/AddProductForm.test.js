import React from 'react';
import AddProductForm from './AddProductForm';
import { shallow } from 'enzyme';

describe('AddProductForm', () => {
  let mockFunc = jest.fn()
  let items = [{name: "Berries",
  id: 1}, {name: "Eggs",
  id: 11}]
  let user = {
    name: "Lemon Sisters",
    email: "regular@email.com",
    address: "12000 E. 47th Ave",
    phone: "(231) 341-4141",
    bio: "We Sell Lemons",
    lat: 39.7820535,
    long: -104.8503416,
    city: "denver",
    state: "CO",
    zip: 80239,
    img_url: null,
    token: "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoyLCJleHAiOjE1NTA3MjM0OTZ9.az6Fs46yi1DL6u9J9cUWf3dt2aSbtsTHCNcTDGAQ-8k",
    products: [{name: "Broccoli",
    id: 9,
    user_id: 2,
    item_id: 2,
    price: 250,
    unit: "lb",
    description: "a lb of broccoli",
    vendorName: "Lemon Sisters"}]
  }
  it('should match snapshot', () => {
    let wrapper = shallow(<AddProductForm postItem={mockFunc} fetchUser={mockFunc} items={items} user={user}/>)
    expect(wrapper).toMatchSnapshot()
  });
})