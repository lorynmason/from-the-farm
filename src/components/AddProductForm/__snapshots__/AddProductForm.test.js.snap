// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProductForm should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(AddProductForm)
    fetchUser={[MockFunction]}
    items={
      Array [
        Object {
          "id": 1,
          "name": "Berries",
        },
        Object {
          "id": 11,
          "name": "Eggs",
        },
      ]
    }
    postItem={[MockFunction]}
    user={
      Object {
        "address": "12000 E. 47th Ave",
        "bio": "We Sell Lemons",
        "city": "denver",
        "email": "regular@email.com",
        "img_url": null,
        "lat": 39.7820535,
        "long": -104.8503416,
        "name": "Lemon Sisters",
        "phone": "(231) 341-4141",
        "products": Array [
          Object {
            "description": "a lb of broccoli",
            "id": 9,
            "item_id": 2,
            "name": "Broccoli",
            "price": 250,
            "unit": "lb",
            "user_id": 2,
            "vendorName": "Lemon Sisters",
          },
        ],
        "state": "CO",
        "token": "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoyLCJleHAiOjE1NTA3MjM0OTZ9.az6Fs46yi1DL6u9J9cUWf3dt2aSbtsTHCNcTDGAQ-8k",
        "zip": 80239,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
